File: python/krpc/test/test_decoder.py
#!/usr/bin/env python2

import unittest
import binascii
from krpc.decoder import _Decoder as Decoder
from krpc.types import _Types as Types
import krpc.schema.KRPC

class TestDecoder(unittest.TestCase):

    def test_decode_message(self):
        typ = krpc.schema.KRPC.Request
        message = '0a0b536572766963654e616d65120d50726f6365647572654e616d65'
        request = Decoder.decode(binascii.unhexlify(message), Types().as_type('KRPC.Request'))
        self.assertEquals('ServiceName', request.service)
        self.assertEquals('ProcedureName', request.procedure)

    def test_decode_value(self):
        value = Decoder.decode(binascii.unhexlify('ac02'), Types().as_type('int32'))
        self.assertEquals(int(300), value)

    def test_decode_size_and_position(self):
        message = '1c'
        size,position = Decoder.decode_size_and_position(binascii.unhexlify(message))
        self.assertEquals(28, size)
        self.assertEquals(1, position)

    def test_decode_message_delimited(self):
        typ = krpc.schema.KRPC.Request
        message = '1c'+'0a0b536572766963654e616d65120d50726f6365647572654e616d65'
        request = Decoder.decode_delimited(binascii.unhexlify(message), Types().as_type('KRPC.Request'))
        self.assertEquals('ServiceName', request.service)
        self.assertEquals('ProcedureName', request.procedure)

    def test_decode_value_delimited(self):
        value = Decoder.decode_delimited(binascii.unhexlify('02'+'ac02'), Types().as_type('int32'))
        self.assertEquals(300, value)

    def test_decode_class(self):
        typ = Types().as_type('Class(ServiceName.ClassName)')
        value = Decoder.decode(binascii.unhexlify('ac02'), typ)
        self.assertTrue(isinstance(value, typ.python_type))
        self.assertEqual(300, value._object_id)

    def test_decode_class_none(self):
        typ = Types().as_type('Class(ServiceName.ClassName)')
        value = Decoder.decode(binascii.unhexlify('00'), typ)
        self.assertIsNone(value)


if __name__ == '__main__':
    unittest.main()

File: client/cpp/test/test_decoder.cpp
#include <gtest/gtest.h>

#include <string>
#include <utility>

#include <krpc/decoder.hpp>
#include <krpc/platform.hpp>

#include "services/test_service.hpp"

namespace pb = google::protobuf;

TEST(test_decoder, test_decode_message) {
  std::string message = "0a0b536572766963654e616d65120d50726f6365647572654e616d65";
  krpc::schema::Request request;
  krpc::decoder::decode(request, krpc::platform::unhexlify(message));
  ASSERT_EQ("ServiceName", request.service());
  ASSERT_EQ("ProcedureName", request.procedure());
}

TEST(test_decoder, test_decode_value) {
  unsigned int value;
  krpc::decoder::decode(value, krpc::platform::unhexlify("ac02"));
  ASSERT_EQ(300, value);
}

TEST(test_decoder, test_decode_unicode_string) {
  std::string value;
  krpc::decoder::decode(value, krpc::platform::unhexlify("03e284a2"));
  ASSERT_EQ(krpc::platform::unhexlify("e284a2"), value);
}

TEST(test_decoder, test_decode_size_and_position) {
  std::string message = "1c";
  std::pair<pb::uint32, pb::uint32> result =
    krpc::decoder::decode_size_and_position(krpc::platform::unhexlify(message));
  ASSERT_EQ(28, result.first);
  ASSERT_EQ(1, result.second);
}

TEST(test_decoder, test_decode_class) {
  krpc::services::TestService::TestClass object;
  krpc::decoder::decode(object, krpc::platform::unhexlify("ac02"));
  ASSERT_EQ(krpc::services::TestService::TestClass(nullptr, 300), object);
}

TEST(test_decoder, test_decode_class_none) {
  krpc::services::TestService::TestClass object;
  krpc::decoder::decode(object, krpc::platform::unhexlify("00"));
  ASSERT_EQ(krpc::services::TestService::TestClass(), object);
}

TEST(test_decoder, test_guid) {
  ASSERT_EQ(
    "6f271b39-00dd-4de4-9732-f0d3a68838df",
    krpc::decoder::guid(krpc::platform::unhexlify("391b276fdd00e44d9732f0d3a68838df")));
}
